File: interactors/src/test/java/fr/coppernic/lib/interactors/robolectric/RobolectricTest.java
package fr.coppernic.lib.interactors.robolectric;

import org.junit.BeforeClass;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.shadows.ShadowLog;

import java.util.concurrent.atomic.AtomicBoolean;

import static junit.framework.TestCase.fail;
import static org.awaitility.Awaitility.await;

/**
 * Base class extended by every Robolectric test in this project.
 * <p>
 * Robolectric tests are done in a single thread !
 */
@RunWith(RobolectricTestRunner.class)
public abstract class RobolectricTest {

    private AtomicBoolean unblock = new AtomicBoolean(false);

    @BeforeClass
    public static void beforeClass() {
        //Configure robolectric
        ShadowLog.stream = System.out;
    }

    public void sleep(long ms) {
        try {
            Thread.sleep(ms);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    public void unblock() {
        unblock.set(true);
    }

    public void block() {
        await().untilTrue(unblock);
    }

    public void doNotGoHere() {
        fail();
    }

}

File: countries/src/test/java/fr/coppernic/lib/test/RobolectricTest.kt
package fr.coppernic.lib.test

import android.os.Build
import fr.bipi.tressence.console.SystemLogTree
import fr.coppernic.lib.countries.log.LogDefines
import org.awaitility.Awaitility.await
import org.junit.AfterClass
import org.junit.BeforeClass
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config
import org.robolectric.shadows.ShadowLog
import timber.log.Timber
import java.util.concurrent.atomic.AtomicBoolean


/**
 * Base class extended by every Robolectric test in this project.
 * <p>
 * Robolectric tests are done in a single thread !
 */
@RunWith(RobolectricTestRunner::class)
@Config(manifest = Config.NONE, sdk = [Build.VERSION_CODES.P])
abstract class RobolectricTest {
    companion object {
        private val tree = SystemLogTree()

        @BeforeClass
        @JvmStatic
        fun beforeClass() {
            Timber.plant(tree)
            ShadowLog.stream = System.out
            LogDefines.setVerbose(true)
        }

        @AfterClass
        @JvmStatic
        fun afterClass() {
            Timber.uproot(tree)
        }
    }

    private val unblock = AtomicBoolean(false)

    fun sleep(ms: Long) {
        try {
            Thread.sleep(ms)
        } catch (e: InterruptedException) {
            e.printStackTrace()
        }

    }

    fun unblock() {
        unblock.set(true)
    }

    fun block() {
        unblock.set(false)
        await().untilTrue(unblock)
    }
}
